api/transport/response.go

```diff
@@ -47,3 +48,14 @@ type ResponseWriter interface {
 	// of Write().
 	SetApplicationError()
 }
+
+// ApplicationErrorNameSetter enables setting the name of an
+// application error, surfacing it in metrics.
+//
+// Conditionally upcast a ResponseWriter to access the
+// functionality.
+type ApplicationErrorNameSetter interface {
```

Comment:
Discussed this with @AlexAPB. ApplicationErrorNameSetter isn't scalable
because there's a likelihood of needing more in the future.

How about an ApplicationFailure object?

```
type ApplicationFailure struct {
  Err  error
  Name string  // optional
}

type ApplicationFailureSetter interface {
  SetApplicationFailure(*ApplicationFailure)
}
```

And above,

```
type Response struct {
  ...
  
  ApplicationError bool
  ApplicationFailure *ApplicationFailure
```

Advantages:

- Adding more fields to ApplicationFailure in the future without introducing
  new interfaces to upcast to.
- SetApplicationFailure implementations can implicitly SetApplicationError,
  allowing a path to deprecate SetApplicationError in favor of
  SetApplicationFailure.
- Access to the raw error leaves room for more metadata on the error object.

Reply:
I agree with this. 

It would also make it easier for you guys to extend it in the future for other encodings. Will update it with this change.